File: leetcode/medium/course_schedule.py
from collections import defaultdict
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for x in range(numCourses)]
        indeg = defaultdict(int)
        # print(graph)
        for x in range(numCourses):
            indeg[x] = 0
        for edge in prerequisites:
            graph[edge[0]].append(edge[1])
            indeg[edge[1]] += 1

        q = []
        for k,v in indeg.items():
            if v == 0: q.append(k)

        c = 0
        while q:
            a = q.pop(0)
            c += 1
            for x in graph[a]:
                indeg[x] -= 1
                if indeg[x] == 0:
                    q.append(x)
        return c == numCourses

File: leetcode/medium/courseSchedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<pair<vector<int>, int>> graph(numCourses);
        for (auto p : prerequisites) {
            cout << p[0] << endl;
            graph[p[1]].first.push_back(p[0]);
            graph[p[0]].second++;
        }
        queue<int> q;
        for (int z = 0; z < numCourses; z++)
            if (!graph[z].second) q.push(z);

        vector<bool> seen(numCourses);
        int count = 0;
        while (!q.empty()) {
            int t = q.front();
            q.pop();
            if (seen[t]) continue;
            seen[t] = true; count++;
            if (count == numCourses) return true;
            for (int a : graph[t].first)
                if (--graph[a].second == 0)
                    q.push(a);
        }
        return count == numCourses;
    }
};
